src/upgradeConfig.js

```diff
@@ -1,110 +1,91 @@
+function appendFlowIfNeeded(presets) {
+  return presets.find(preset => preset === '@babel/preset-flow' || preset[0] === '@babel/preset-flow')
+    ? presets
+    : presets.concat(['@babel/preset-flow']);
+}
+
+function transformPreset(preset) {
+  let presetName = Array.isArray(preset) ? preset[0] : preset;
+  const presetOptions = Array.isArray(preset) ? preset[1] : {};
+
+  if (presetName.indexOf('babel-preset') !== 0 && presetName.indexOf('@babel/') !== 0) {
+    presetName = `babel-preset-${presetName}`;
+  }
 
+  const newPresetName = oldPresets[presetName];
+  if (newPresetName !== undefined) {
+    return Object.keys(presetOptions).length > 0 ? [newPresetName, presetOptions] : newPresetName;
+  } else {
+    return preset;
+  }
+
+  return null;
+}
+
+function transformPlugins(plugins) {
   if (!Array.isArray(plugins) && typeof plugins === 'string') {
-    plugins = config.plugins = config.plugins.split(',').map((plugin) => plugin.trim());
+    plugins = plugins.split(',').map(plugin => plugin.trim());
   }
 
-  // check if plugins are there
-  if (plugins) {
-    // assume it's an array
-    for (let i = 0; i < plugins.length; i++) {
-      let plugin = plugins[i];
-      const pluginsToReplace = Object.keys(oldPlugins);
-
-      // check if it's a plugin with options (an array)
-      if (Array.isArray(plugin)) {
-
-        if (plugin[0].indexOf('babel-plugin') !== 0 && plugin[0].indexOf('@babel/') !== 0) {
-          plugin[0] = `babel-plugin-${plugin[0]}`;
-        }
-
-        if (pluginsToReplace.includes(plugin[0])) {
-          if (oldPlugins[plugin[0]]) {
-            plugin[0] = oldPlugins[plugin[0]];
-          } else {
-            plugins.splice(i, 1);
-            i--;
-          }
-        }
-      } else {
-        if (plugin.indexOf('babel-plugin') !== 0 && plugin[0].indexOf('@babel/') !== 0) {
-          plugin = `babel-plugin-${plugin}`;
-        }
-        if (pluginsToReplace.includes(plugin)) {
-          if (oldPlugins[plugin]) {
-            plugins[i] = oldPlugins[plugin];
-          } else {
-            plugins.splice(i, 1);
-            i--;
-          }
-        }
-      }
-    }
+  return plugins && plugins.map(transformPlugin).filter(Boolean);
+}
+
+function transformPlugin(plugin) {
+  let pluginName = Array.isArray(plugin) ? plugin[0] : plugin;
+  const pluginOptions = Array.isArray(plugin) ? plugin[1] : {};
+
+  if (pluginName.indexOf('babel-plugin') !== 0 && pluginName.indexOf('@babel/') !== 0) {
+    pluginName = `babel-plugin-${pluginName}`;
+  }
+
+  const newPluginName = oldPlugins[pluginName];
+  if (newPluginName !== undefined) {
+    return Object.keys(pluginOptions).length > 0 ? [newPluginName, pluginOptions] : newPluginName;
+  } else {
+    return plugin;
   }
+
+  return null;
 }
 
 module.exports = function upgradeConfig(config, options) {
   config = Object.assign({}, config);
 
-  changePresets(config, options);
-  changePlugins(config);
+  if (config.presets) {
```

Comment:
i was thinking it might be clearer to only mutate config within `upgradeConfig`, but that might just be personal preference